proposals/NNNN-high-level-interface-modal-view-controllers.md

```diff
@@ -0,0 +1,160 @@
+
+The oldest VC pair of this list should be the currently presenting and presented VC. The currently presenting VC will be accessible through the `.presentingModalViewController` property and the presented VC through the `.presentedModalViewController`. If the private list is empty the two properties should return nil.
+
+By the nature of the HMI overlays can be aborted by another overlay or get rejected because of a currently presented overlay (or something of a higher priority e.g. VR session). The private list offers flexibility in case the app calls the method multiple times and can deal with rejected overlays that were never visible and aborted overlays if another one is requested.
+
+#### SDLModalViewController
+
+This class should be used as the base class for any kind of overlay. As it inherits `SDLViewController` it is possible to manipulate the head unit during a modal presentation.
+
+```objc
+@interface SDLModalViewController : SDLViewController
+
+@property (nonatomic, assign) NSTimeInterval duration; // used throughout all subclasses
+@property (nonatomic, copy) NSString *message; // used throughout all subclasses
+
+- (void)present:(SDLApplication *)application; // private method called by the VC manager
+
+@end
+```
+
+#### SDLAlertController
+
+The alert controller is used to abstract the RPC `Alert`. The properties match the parameters of the RPC.
+
+```objc
+@interface SDLAlertController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
+
+@property (nonatomic) BOOL shouldPlayTone;
+@property (nonatomic) BOOL progressIndicator;
+
+@end
+``` 
+
+#### SDLAudioInputController
+
+The audio input controller is used to abstract the RPC `PerformAudioPassThru`, `OnAudioPassThru` and `EndAudioPassThru`. The app developer should be able to provide a block which is called for audio data. Furthermore a method called `dismiss` should help to end the audio pass thru.
+
+```objc
+@interface SDLAudioInputController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLTTSChunk *> *initialPrompt;
+@property (nonatomic, copy) SDLSamplingRate samplingRate;
+@property (nonatomic, copy) SDLBitsPerSample bitsPerSample;
+@property (nonatomic, copy) SDLAudioType audioType;
+@property (nonatomic) BOOL muteMediaSource;
+
+@property (nonatomic) (^onAudioData)(NSData *audioData);
+
+- (void)finish; // sends an EndAudioPassThru in order to stop audio input
+
+@end
+```
+
+#### SDLScrollableMessageController
+
+Similar to `Alert` this view controller takes care of the RPC `ScrollableMessage`.
+
+```objc
+@interface SDLScrollableMessageController : SDLModalViewController
+
+@property (nonatomic, copy) NSArray<SDLSoftButton *> *buttons;
```

Comment:
Needs a body property

Reply:
These should probably be readonly and created in the setter. If they change them during the presentation, things could get weird or unexpected for the developer

Reply:
Actually it doesn't need a body. `SDLModalViewController.message` will be used.

Reply:
I think that was a criticism of mine on the first proposal. If additional modals are created that _don't_ have a message, then having a `message` property on `SDLModalViewController` becomes a problem. I am personally strongly in favor of removing `message` from there and putting it on each subclass.